internal/controllers/user_devices_controller.go

```diff
@@ -1856,6 +1859,60 @@ func (udc *UserDevicesController) PostMintDevice(c *fiber.Ctx) error {
 	}, sigBytes)
 }
 
+// GetUserDevicesWithCredentials godoc
+// @Description Complete Tesla auth and get devices for authenticated user
+// @Tags        user-devices
+// @Produce     json
+// @Accept      json
+// @Param       user_device body controllers.GetUserDevicesByIntegrationRequest true "all fields are required"
+// @Security    ApiKeyAuth
+// @Success     200 {object} controllers.GetUserDevicesByIntegrationResponse
+// @Security    BearerAuth
+// @Router      /user/devices/{tokenID}/credentials [post]
+func (udc *UserDevicesController) GetUserDevicesWithCredentials(c *fiber.Ctx) error {
+	reqBody := new(GetUserDevicesWithCredentialsRequest)
+	if err := c.BodyParser(reqBody); err != nil {
+		return fiber.NewError(fiber.StatusBadRequest, "Couldn't parse request JSON body.")
+	}
+
+	if reqBody.Region != "na" && reqBody.Region != "eu" {
+		return fiber.NewError(fiber.StatusBadRequest, "invalid value provided for region, only na and eu are allowed")
+	}
+
+	logger := udc.log.With().
+		Str("region", reqBody.Region).
+		Str("redirectURI", reqBody.RedirectURI).
+		Str("route", c.Route().Path).
+		Logger()
+	logger.Info().Msg("Attempting to complete Tesla authorization")
+
+	teslaAuth, err := udc.teslaAPIService.CompleteTeslaAuthCodeExchange(reqBody.AuthorizationCode, reqBody.RedirectURI, reqBody.Region)
+	if err != nil {
+		return err
+	}
+
+	vehicles, err := udc.teslaAPIService.GetVehicles(teslaAuth.AccessToken, reqBody.Region)
+	if err != nil {
+		return err
+	}
+
+	response := make([]GetUserDevicesWithCredentialsResponse, 0, len(vehicles))
+
+	for _, v := range vehicles {
+		response = append(response, GetUserDevicesWithCredentialsResponse{
```

Comment:
Make `response[i] = ...`?

Reply:
For this remember to change `response := make([]GetUserDevicesWithCredentialsResponse, 0, len(vehicles))` => `response := make([]GetUserDevicesWithCredentialsResponse, len(vehicles))` once compiled I believe both have the same performance since the underlying array is allocated to the correct size. I usually lean towards `response[i] = v` but that is just for stylistic reasons.